src/core/createPackageLocator.js

```diff
@@ -0,0 +1,69 @@
+import path from 'path'
+import fs from 'fs'
+import { EOL } from 'os'
+
+class PackageLocator {
+  readPackageSync(location) {
+    return {
+      result: JSON.parse(fs.readFileSync(location, 'utf8')),
+      location,
+    }
+  }
+
+  readPackageUpSync(dirname) {
+    const locations = []
+
+    do {
+      try {
+        const location = path.join(dirname, 'package.json')
+        locations.push(location)
+
+        return this.readPackageSync(location)
+      } catch (err) {
+        // if not found, allopw search to continue
+        if (err.code !== 'ENOENT') throw err
+      }
+    } while (dirname !== (dirname = path.dirname(dirname)))
```

Comment:
is there a way this can be written without conflating an assignment with an expression?

Reply:
@ljharb Yes, but this plays nicer with the goal of "fail fast". Meaning, if i were to replace with a typical condition-first loop such as for/while - it would require me to add logic which ensured the initial dirname was searched.